src/safeposix/dispatcher.rs

```diff
@@ -201,252 +176,671 @@ pub extern "C" fn rustposix_thread_init(cageid: u64, signalflag: u64) {
     interface::signalflag_set(signalflag);
 }
 
+/// The `lind_syscall_api` function acts as the main dispatcher for handling system calls 
+/// within the Lind virtualized environment. It identifies the syscall to execute based on 
+/// `call_number`, and then invokes the appropriate syscall with the given arguments within 
+/// the specified cage (`cageid`).
+/// 
+/// ### Arguments:
+/// `lind_syscall_api()` accepts 10 arguments:
+/// * `cageid` - Identifier for the cage in which the syscall will be executed.
+/// * `call_number` - Unique number for each system call, used to identify which syscall to invoke.
+/// * `call_name` - A legacy argument from the initial 3i proposal, currently unused and subject to 
+///                 change with future 3i integration.
+/// * `start_address` - Base address of WebAssembly linear memory, used for address translation
+///                     between virtual and system memory address.
+/// * `arg1 - arg6` - Syscall-specific arguments. Any unused argument is set to `0xdeadbeefdeadbeef`.
+/// 
+/// ### Returns:
+/// On success, returns the syscall's return value. On failure, returns the negative errno code.
+/// 
+/// ### Panics:
+/// * If the specified `cageid` does not exist, the function will panic.
 #[no_mangle]
-pub extern "C" fn dispatcher(
+pub fn lind_syscall_api(
     cageid: u64,
-    callnum: i32,
-    arg1: Arg,
-    arg2: Arg,
-    arg3: Arg,
-    arg4: Arg,
-    arg5: Arg,
-    arg6: Arg,
+    call_number: u32,
+    call_name: u64,
+    start_address: u64,
+    arg1: u64,
+    arg2: u64,
+    arg3: u64,
+    arg4: u64,
+    arg5: u64,
+    arg6: u64,
 ) -> i32 {
-    // need to match based on if cage exists
-    let cage = interface::cagetable_getref(cageid);
+    let call_number = call_number as i32;
+
+    let ret = match call_number {
+        WRITE_SYSCALL => {
+            let fd = arg1 as i32;
+            let buf = (start_address + arg2) as *const u8;
+            let count = arg3 as usize;
+            interface::check_cageid(cageid);
+            unsafe {
```

Comment:
I didn't realize we were using unsafe to do all these calls, I think this is probably bad. Why do we do this?